File: utbot-framework/src/main/java/org/utbot/quickcheck/generator/java/util/function/BiPredicateGenerator.java


package org.utbot.quickcheck.generator.java.util.function;
import org.utbot.engine.greyboxfuzzer.util.UtModelGenerator;
import org.utbot.framework.plugin.api.UtModel;

import org.utbot.quickcheck.generator.ComponentizedGenerator;
import org.utbot.quickcheck.generator.GenerationStatus;
import org.utbot.quickcheck.generator.Generator;
import org.utbot.quickcheck.generator.Generators;
import org.utbot.quickcheck.random.SourceOfRandomness;

import java.util.function.BiPredicate;

import static org.utbot.external.api.UtModelFactoryKt.classIdForType;
import static org.utbot.quickcheck.generator.Lambdas.makeLambda;

/**
 * Produces values of type {@link BiPredicate}.
 *
 * @param <T> type of first parameter of produced predicate
 * @param <U> type of second parameter of produced predicate
 */
public class BiPredicateGenerator<T, U>
    extends ComponentizedGenerator<BiPredicate> {

    private Generator<Boolean> generator;

    public BiPredicateGenerator() {
        super(BiPredicate.class);
    }

    @Override public void provide(Generators provided) {
        super.provide(provided);

        generator = gen().type(boolean.class);
    }

    @Override public UtModel generate(
        SourceOfRandomness random,
        GenerationStatus status) {

        return UtModelGenerator.getUtModelConstructor().construct(makeLambda(BiPredicate.class, generator, status), classIdForType(BiPredicateGenerator.class));
    }

    @Override public int numberOfNeededComponents() {
        return 2;
    }
}

File: utbot-framework/src/main/kotlin/org/utbot/quickcheck/generator/java/util/function/BiPredicateGenerator.kt
package org.utbot.quickcheck.generator.java.util.function

import org.utbot.engine.greyboxfuzzer.util.UtModelGenerator.utModelConstructor
import org.utbot.external.api.classIdForType
import org.utbot.framework.plugin.api.UtModel
import org.utbot.quickcheck.generator.ComponentizedGenerator
import org.utbot.quickcheck.generator.GenerationStatus
import org.utbot.quickcheck.generator.Generator
import org.utbot.quickcheck.generator.Generators
import org.utbot.quickcheck.generator.Lambdas.Companion.makeLambda
import org.utbot.quickcheck.random.SourceOfRandomness
import java.util.function.BiPredicate

/**
 * Produces values of type [BiPredicate].
 *
 * @param <T> type of first parameter of produced predicate
 * @param <U> type of second parameter of produced predicate
</U></T> */
class BiPredicateGenerator<T, U> : ComponentizedGenerator(BiPredicate::class.java) {
    private var generator: Generator? = null

    override fun createModifiedUtModel(random: SourceOfRandomness, status: GenerationStatus): UtModel {
        return generate(random, status)
    }

    override fun provide(provided: Generators) {
        super.provide(provided)
        generator = gen()!!.type(Boolean::class.javaPrimitiveType!!)
    }

    override fun generate(
        random: SourceOfRandomness,
        status: GenerationStatus
    ): UtModel {
        return utModelConstructor.construct(
            makeLambda(BiPredicate::class.java, generator!!, status),
            classIdForType(BiPredicate::class.java)
        )
    }

    override fun numberOfNeededComponents(): Int {
        return 2
    }
}